### PR TITLE
Allow relcast to serialize raw erlang terms

### PR DESCRIPTION
**Pull request by Vagabond**
The goal of this change is to move any needed t2b serialization as far to the edge as possible, so we avoid unnecessary term_to_binary calls when fields don't change AND we also retain binary references at the keytree compare phase. Binary references are important because they allow a much faster equality check for large binaries than a byte-by-byte compare.

### CODE REVIEW

File: src/relcast.erl [hunk 1]
```diff
@@ -1224,19 +1224,31 @@ do_serialize(Mod, Old, New, Prefix, Transaction) ->
...
+                              false when is_binary(V) ->
                                   ok = rocksdb:transaction_put(Transaction, KeyName, V),
+                                  K;
+                              false ->
+                                  %% some other term, try to term to binary as close to the edge
+                                  %% as possible to avoid the relcast modules
+                                  %% creating large binaries that have to be compared byte by byte
+                                  Encoded = term_to_binary(V, [compressed]),
+                                  %% store it tagged so we know to b2t it on deserialize
+                                  ok = rocksdb:transaction_put(Transaction, KeyName, <<"relcast-", Encoded/binary>>),
```
Comment (Vagabond) [inline src/relcast.erl:1239]: Could use better tag bytes here.

File: src/relcast.erl [hunk 2]
```diff
@@ -1224,19 +1224,31 @@ do_serialize(Mod, Old, New, Prefix, Transaction) ->
...
-                              false ->
-                                  %% lager:info("writing ~p to disk", [K]),
+                              false when is_binary(V) ->
                                   ok = rocksdb:transaction_put(Transaction, KeyName, V),
+                                  K;
+                              false ->
+                                  %% some other term, try to term to binary as close to the edge
+                                  %% as possible to avoid the relcast modules
+                                  %% creating large binaries that have to be compared byte by byte
+                                  Encoded = term_to_binary(V, [compressed]),
```
Comment (Vagabond) [inline src/relcast.erl:1237]: We should probably make the t2b options come from the app env, so they could be configured.

File: src/relcast.erl [hunk 3]
```diff
@@ -1224,19 +1224,31 @@ do_serialize(Mod, Old, New, Prefix, Transaction) ->
...
+                                  ok = rocksdb:transaction_put(Transaction, KeyName, <<"relcast-", Encoded/binary>>),
                                   K
                           end
                   end,
                   L),
             [Mod | KeyTree]
     end.
 
+is_keytree_map(M) when is_map(M) ->
+    lists:all(fun is_atom/1, maps:keys(M));
```
[the remark this answers is not in the document]
Comment (marcsugiyama) [reply src/relcast.erl:1248]: There are maps in some states with non-atom keys, but these will fall through to the third case cause.

[the remark this answers is not in the document]
Comment (marcsugiyama) [reply src/relcast.erl:1248]: But these may turned into binary in the serializers. For example `hbbft:serialize_hbbft_data/1`. I think the keys of the maps are not atoms.
This function is calling `term_to_binary` with `compressed` on most of the keys. Can we avoid this?

**Review by marcsugiyama — APPROVED** [review]

File: src/relcast.erl (hunk 2 above)
Comment (evanmcc) [reply src/relcast.erl:1237]: done

**Review by evanmcc — APPROVED** [review]

**Pull request merged**